### PR TITLE
BUG: Wrap std::sto* exceptions in IO with itk::ExceptionObject (#3213)

### LINKED ISSUE #3213: Wrap lowlevel c++ string conversion exceptions with ITK exceptions.
### Description

ITK was recently changed from using 'ato[ifd]' to using 'sto[ifd]' in an attempt to provide more robust error checking.

```
commit b9a384af33e904ae0da212ace9130632c808c6fd
Author: Hans Johnson <hans-johnson@uiowa.edu>
Date:   Sat Sep 22 12:57:23 2018

    STYLE: Prefer error checked std::stoi over atoi
    
    The atoi function does not provide mechanisms for distinguishing between
    '0' and the error condion where the input can not be converted.
    
    std::stoi provides exception handling and detects when an invalid
    string attempts to be converted to an integer.
    
    atoi()
      Con: No error handling.
      Con: Handle neither hexadecimal nor octal.
    
    The use of atoi in code can cause it to be subtly broken.
    atoi makes two very big assumptions indeed:
      The string represents an integer/floating point value.
      The integer can fit into an int.
```

### Impact analysis

The above change now introduces a data dependant throw of low-level C++ exceptions.  ITK should convert these into ITK exceptions instead, with non-generic messages based on the situation.

### Expected behavior

ITK generates context-specific ITK exceptions in data-induced exceptions.

### Actual behavior

Default low level c++ exeptions are thrown of type
```
If no conversion could be performed, an [invalid_argument](https://www.cplusplus.com/invalid_argument) exception is thrown.

If the value read is out of the range of representable values by an int, an [out_of_range](https://www.cplusplus.com/out_of_range) exception is thrown.
```

### Versions

All TIK versions after commit b9a384af33e904ae0da212ace9130632c808c6fd

### Additional Information

This change was discussed in the context of : https://github.com/Slicer/Slicer/pull/6200

### PR DESCRIPTION
**Pull request by hjmjohnson**
Fixes #3213 by introducing context-aware `itk::StringTo*` helpers and migrating non-test `std::sto*` call sites in the IO modules to use them. Corrupt input now raises an `itk::ExceptionObject` (with the offending field name and value) rather than a bare `std::invalid_argument` / `std::out_of_range` that crashes downstream applications such as Slicer.

<details>
<summary>Background and rationale</summary>

ITK switched from atoi/atof to `std::stoi` / `std::stod` in commit b9a384af3 (2018) for better error checking. As discussed in Slicer/Slicer#6200, that change introduced a behavior regression for callers that catch only `itk::ExceptionObject`: a corrupt header that previously parsed silently as zero now throws an exception unrelated to ITK's exception hierarchy and crashes the application. The Slicer reviewers asked that ITK catch the low-level conversion exceptions internally and rethrow them as `itk::ExceptionObject` with a meaningful message — which is what this PR does.

Behavior tiers as framed in that thread:
- WORST: silently produce wrong results
- REALLY BAD: crash with an unhandled std exception
- BEST: throw a typed contextual exception that downstream catchers already handle

Wrapping moves the IO read paths into the BEST tier without changing success-path behavior.

</details>

<details>
<summary>What changed</summary>

**Commit 1 — ENH: Add itk::StringTo{Int,Long,ULong,Double,Float} helpers**
- New header `Modules/Core/Common/include/itkStringTools.h` and implementation in `src/itkStringTools.cxx`.
- Each helper takes the input string and a `const char * context`. On failure it throws `itk::ExceptionObject` whose message includes the target type, the failure kind (`invalid_argument` or `out_of_range`), the context, the offending input (truncated past 64 chars), and the underlying `what()`.
- `StringToULong` additionally rejects an explicit leading minus sign; `std::stoul` silently wraps it.
- New GoogleTest `itkStringToolsGTest.cxx` covers valid round-trip, empty input, non-numeric input, out-of-range overflow, message content, the long-input truncation path, the null-context path, and the explicit-minus rejection.

**Commit 2 — BUG: Wrap std::sto* call sites in IO with itk::ExceptionObject**

Per-module migration:

- Core/Common — itkSmapsFileParser bracketed size
- IO/Bruker — VisuFGOrderDesc size entries
- IO/GDCM — DICOM (0028,0100/0101/0102/0103)
- IO/GE — SIGNA series and image numbers, images-per-slice
- IO/ImageBase — RegularExpressionSeriesFileNames numeric sort key
- IO/MeshVTK — VTK PolyData header version major
- IO/NIFTI — metadata qform_code and sform_code
- IO/Siemens — Vision header TEXT_* numeric fields
- IO/SpatialObjects — PolygonGroup XML reader
- Video/IO — VideoIOFactory camera index

Plus a new corruption test `itkPolygonGroupXMLMalformedTest` that feeds malformed XML and asserts the failure surfaces as `itk::ExceptionObject` (not `std::exception`), locking in the wrapping contract end-to-end for at least one IO module.

</details>

<details>
<summary>Scope notes (what is intentionally NOT touched)</summary>

- **Test-side `std::sto*` calls** (parsing argv in CTest drivers; roughly 60 sites) are intentionally left as-is. A stack trace from a developer-supplied bad command-line argument is more diagnostic than a wrapped exception; the Slicer reviewers explicitly noted that test-side crashes are desirable.
- **No silent-recovery substitution.** Where a caller legitimately wants empty-string-means-zero behavior (the old atoi semantics) that is the caller's policy and should be expressed as an explicit empty-string check before calling the helper. The Siemens reader already does this for `text_angle_len`; this PR preserves that pattern.
- **ThirdParty trees and Examples are untouched.**

</details>

## Test plan

- [ ] CI green across configured platforms
- [ ] `ITKCommonGTestDriver --gtest_filter=StringTools.*` passes
- [ ] `itkPolygonGroupXMLMalformedTest` passes (asserts wrapped exception)
- [ ] Existing PolygonGroup / NIfTI / GDCM read tests still pass on valid input

### CODE REVIEW

File: Modules/Core/Common/src/itkStringTools.cxx [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+#include <string>
+
+namespace itk
+{
+
+namespace
+{
+// Cap quoted-input length in exception messages so a multi-megabyte
+// header value cannot blow up the exception text.
+constexpr std::size_t maxQuotedInputLength = 64;
```
Comment (hjmjohnson) [inline Modules/Core/Common/src/itkStringTools.cxx]: Change maxQuotedInputLength to 512 characters to provide headroom for better diagnostics.

Comment (hjmjohnson) [reply Modules/Core/Common/src/itkStringTools.cxx]: ensure that there is a test that verifies that maxQuotedInputLength is working as expected.

Comment (hjmjohnson) [reply Modules/Core/Common/src/itkStringTools.cxx]: Done in 47aedd9 — bumped `maxQuotedInputLength` from 64 to 512 for headroom.

Comment (hjmjohnson) [reply Modules/Core/Common/src/itkStringTools.cxx]: Added `StringToInt_TruncationLimitIs512Chars` in 47aedd9 — verifies (a) at-limit input is quoted in full, (b) over-limit input shows the truncation marker with the original byte count, (c) the quoted prefix is exactly 512 sentinel chars and one char beyond the limit does not appear.

File: Modules/IO/ImageBase/src/itkRegularExpressionSeriesFileNames.cxx [hunk 2]
```diff
@@ -22,13 +22,15 @@
...
 struct lt_pair_numeric_string_string
 {
   bool
   operator()(const std::pair<std::string, std::string> & s1, const std::pair<std::string, std::string> & s2) const
   {
-    return std::stod(s1.second.c_str()) < std::stod(s2.second.c_str());
+    return itk::StringToDouble(s1.second, "RegularExpressionSeriesFileNames numeric sort key") <
+           itk::StringToDouble(s2.second, "RegularExpressionSeriesFileNames numeric sort key");
   }
 };
```
[the remark this answers is not in the document]
Comment (hjmjohnson) [reply Modules/IO/ImageBase/src/itkRegularExpressionSeriesFileNames.cxx:35]: Addressed in 8667009 (separate `COMP:` commit). Pre-validates every numeric sort key with `itk::StringToDouble` before `std::sort` is entered, so a malformed value now throws `itk::ExceptionObject` cleanly without entering the comparator-throws-inside-sort UB territory.

File: Modules/IO/SpatialObjects/src/itkPolygonGroupSpatialObjectXMLFile.cxx [hunk 3]
```diff
@@ -79,37 +80,37 @@ PolygonGroupSpatialObjectXMLFileReader::EndElement(const char * name)
...
   {
-    const float res = std::stod(m_CurCharacterData.c_str());
+    const float res = itk::StringToFloat(m_CurCharacterData, "PolygonGroup XML element Y-RESOLUTION");
     itk::EncapsulateMetaData<float>(thisDic, ROI_Y_RESOLUTION, res);
   }
   else if (itksys::SystemTools::Strucmp(name, "Z-RESOLUTION") == 0)
   {
-    const float res = std::stod(m_CurCharacterData.c_str());
+    const float res = itk::StringToFloat(m_CurCharacterData, "PolygonGroup XML element Z-RESOLUTION");
     itk::EncapsulateMetaData<float>(thisDic, ROI_Z_RESOLUTION, res);
```
[the remark this answers is not in the document]
Comment (hjmjohnson) [reply Modules/IO/SpatialObjects/src/itkPolygonGroupSpatialObjectXMLFile.cxx:116]: Behavior change is intentional — see code comment added in 504d0bb. The new `out_of_range` `itk::ExceptionObject` for resolutions in the (FLT_MAX, DBL_MAX] range is the desired strict behavior; silently coercing to `INFINITY` was the bug.

File: Modules/Video/IO/src/itkVideoIOFactory.cxx [hunk 4]
```diff
@@ -57,7 +58,7 @@ VideoIOFactory::CreateVideoIO(IOModeEnum mode, const char * arg)
     // Check camera readability if reading from camera
     else if (mode == IOModeEnum::ReadCameraMode)
     {
-      const int cameraIndex = std::stoi(arg);
+      const int cameraIndex = itk::StringToInt(arg, "VideoIOFactory camera index");
```
[the remark this answers is not in the document]
Comment (hjmjohnson) [reply Modules/Video/IO/src/itkVideoIOFactory.cxx]: Hoisted in 504d0bb — `cameraIndex` is now parsed once before the loop and a malformed argv argument throws independently of how many factories are registered.

File: Modules/Core/Common/test/itkStringConvertGTest.cxx [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  // std::stoul silently wraps "-1"; our wrapper must throw.
+  EXPECT_THROW(itk::StringToULong("-1", "ctx"), itk::ExceptionObject);
+  EXPECT_THROW(itk::StringToULong("  -7", "ctx"), itk::ExceptionObject);
+}
+
+TEST(StringTools, StringToULong_Bad_Throws)
+{
+  EXPECT_THROW(itk::StringToULong("", "ctx"), itk::ExceptionObject);
+  EXPECT_THROW(itk::StringToULong("xyz", "ctx"), itk::ExceptionObject);
+}
```
[the remark this answers is not in the document]
Comment (hjmjohnson) [reply Modules/Core/Common/test/itkStringConvertGTest.cxx]: Added `StringToULong_OutOfRange_Throws` in d98ce34 mirroring the signed test.

**Review by dzenanz — APPROVED** [review]
The fix looks good on a glance. I did not take a good look at the tests.

File: Modules/Core/Common/include/itkStringConvert.h [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+ *  failure, mentioning `context` and the offending input. */
+ITKCommon_EXPORT long
+StringToLong(const std::string & str, const char * context);
+
+/** Parse `str` as a decimal `unsigned long`. Throws itk::ExceptionObject
+ *  on failure, mentioning `context` and the offending input.
+ *  In addition to the underlying `std::stoul` failure modes, an explicit
+ *  leading minus sign is rejected (rather than silently wrapping). */
+ITKCommon_EXPORT unsigned long
+StringToULong(const std::string & str, const char * context);
```
Comment (N-Dekker) [inline Modules/Core/Common/include/itkStringConvert.h]: Do we really still need conversion to `long` and `unsigned long`? You know, `long` and `unsigned long` are problematic because their size is different from one (supported) platform to another. On Windows/Visual C++, long is 32-bits, on Linux, it's typically 64-bits.

Preferably use `long long`, if it must be really long! Otherwise `std::int64_t`.

**Comment by hjmjohnson** [comment]
Renamed `itk::StringToULong` → `itk::StringToUInt64` (returns `std::uint64_t`, internally `std::stoull`). Force-pushed as `3ad9e63435`.

**Why:** `unsigned long` is **32 bits on 64-bit Windows (LLP64)** and 64 bits on Linux/macOS (LP64). For an IO helper that may parse header-field sizes (e.g., NIfTI `> 4 GB`), `unsigned long` would silently throw `out_of_range` on Windows for values that fit comfortably in 64 bits everywhere else. `std::uint64_t` is exactly 64 bits on every platform.

**Internally:** `std::stoull` returns `unsigned long long` (≥ 64 bits per the standard); cast to `std::uint64_t` for fixed-width semantics.

**Tests added:** explicit `9999999999` (> 2^32) round-trip case to lock in the cross-platform width contract; `2^64 = 18446744073709551616` overflow case mirroring the signed test.

No callers outside this PR's own GTest, so the rename is zero-blast-radius. Pre-commit passes locally; CI re-running on the new HEAD.

**Comment by hjmjohnson** [comment]
Refactored to consistent fixed-width naming as suggested. Force-pushed as `bc3ee656f9`.

**API now:**
- `std::int32_t  itk::StringToInt32(...)`
- `std::int64_t  itk::StringToInt64(...)`
- `std::uint32_t itk::StringToUInt32(...)`
- `std::uint64_t itk::StringToUInt64(...)`
- `double        itk::StringToDouble(...)` *(unchanged)*
- `float         itk::StringToFloat(...)` *(unchanged)*

Same width-portability concern as `unsigned long` applied to `long` (LP64 vs LLP64 — `long` is 32-bit on Windows x64, 64-bit on Linux/macOS), so the rename was extended.

**Implementation:**
- `StringToInt32`/`StringToUInt32` use `std::stoll`/`std::stoull` and **explicitly range-check** against `numeric_limits<int32_t>::min/max` / `numeric_limits<uint32_t>::max` before casting. The check fires before the cast, so a value that fits in the underlying `long long` but not in the named width throws `out_of_range` with a clear message rather than truncating.
- `StringToInt64`/`StringToUInt64` use `std::stoll`/`std::stoull` directly (≥ 64-bit guaranteed by the standard) and cast to fixed-width.
- The leading-minus-sign guard for unsigned helpers is factored into a single `rejectLeadingMinus()` helper used by both `UInt32` and `UInt64`.

**Tests:** all six functions get parity coverage — valid round-trip including boundary values (`INT32_MAX`, `INT64_MAX`, `UINT32_MAX`, `UINT64_MAX`), empty / non-numeric / out-of-range, message content, leading-minus rejection for unsigned, and the 512-char truncation contract.

**IO consumers updated:** `itk::StringToInt(` → `itk::StringToInt32(` in 9 files (Smaps, Bruker, GDCM, GE, ImageBase, MeshVTK, NIFTI, Siemens, SpatialObjects, Video/IO). All call sites previously cast the `int` result to `unsigned short` / `SizeValueType` etc., which is unchanged.

Pre-commit clean locally; CI re-running on the new HEAD.

File: Modules/Core/Common/include/itkStringConvert.h (hunk 6 above)
Comment (hjmjohnson) [reply Modules/Core/Common/include/itkStringConvert.h]: Exactly right — addressed in 8600c87 (force-pushed). The API now uses fixed-width `std::int32_t`/`int64_t`/`uint32_t`/`uint64_t` throughout: `StringToLong` and `StringToULong` are gone; the helpers are now `StringToInt32`/`StringToInt64`/`StringToUInt32`/`StringToUInt64`. Internally they call `std::stoll`/`std::stoull` (≥ 64-bit guaranteed) and explicit-range-check before casting to the narrower fixed-width types where applicable. See https://github.com/InsightSoftwareConsortium/ITK/pull/6089#issuecomment-4283389702 for the full rationale and test coverage.

**Comment by hjmjohnson** [comment]
@N-Dekker Thank you for catching this.  It is an improvement over the historical code!

File: Modules/Core/Common/src/itkStringConvert.cxx [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+{
+
+namespace
+{
+// Cap quoted-input length in exception messages so a multi-megabyte
+// header value cannot blow up the exception text.
+constexpr std::size_t maxQuotedInputLength = 512;
+
+std::string
+quoteForMessage(const std::string & str)
```
Comment (N-Dekker) [inline Modules/Core/Common/src/itkStringConvert.cxx]: Preferably use a capital initial letter for function names, following ITK's naming conversion.

File: Modules/Core/Common/src/itkStringConvert.cxx [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+void
+rejectLeadingMinus(const char * context, const std::string & str, const char * targetType)
+{
+  for (const char c : str)
+  {
+    if (c == '-')
+    {
+      throwParseFailure(context, str, targetType, "leading minus sign", "invalid_argument");
+    }
+    if (c != ' ' && c != '\t' && c != '\n' && c != '\r' && c != '\f' && c != '\v' && c != '+')
```
Comment (N-Dekker) [inline Modules/Core/Common/src/itkStringConvert.cxx:68]: Are such whitespace characters valid, in a string that represents a number?

File: Modules/Core/Common/src/itkStringConvert.cxx [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+// Cap quoted-input length in exception messages so a multi-megabyte
+// header value cannot blow up the exception text.
+constexpr std::size_t maxQuotedInputLength = 512;
+
+std::string
+quoteForMessage(const std::string & str)
+{
+  if (str.size() <= maxQuotedInputLength)
+  {
+    return std::string{ "'" } + str + "'";
```
Comment (N-Dekker) [inline Modules/Core/Common/src/itkStringConvert.cxx]: Why not simply `return "'" + str + "'"`?

File: Modules/Core/Common/src/itkStringConvert.cxx (hunk 7 above)
Comment (hjmjohnson) [reply Modules/Core/Common/src/itkStringConvert.cxx]: Addressed in `25b3ea2b` — renamed the file-local helpers to ITK CamelCase: `quoteForMessage` → `QuoteForMessage`, `throwParseFailure` → `ThrowParseFailure`, `rejectLeadingMinus` → `RejectLeadingMinus`.

File: Modules/Core/Common/src/itkStringConvert.cxx (hunk 9 above)
Comment (hjmjohnson) [reply Modules/Core/Common/src/itkStringConvert.cxx]: Addressed in `25b3ea2b` — simplified to `"'" + str + "'"` (and the same pattern in the truncated branch). The earlier `std::string{"'"} + ...` was defensive; `const char* + std::string` is well-defined by the `std::operator+` overloads in `<string>`.

File: Modules/Core/Common/src/itkStringConvert.cxx (hunk 8 above)
Comment (hjmjohnson) [reply Modules/Core/Common/src/itkStringConvert.cxx:68]: Intentional — matches `std::strtoul`'s leading-whitespace behavior per the C standard (`std::isspace` on the "C" locale accepts space, tab, newline, CR, form-feed, and vertical-tab). Since `StringTo*` is a thin wrapper over `std::sto*` which itself calls through to `strtoul` family, rejecting leading whitespace here would create a stricter-than-underlying-C contract and surprise callers that already depend on the C-standard skip behavior (e.g. whitespace-delimited tokens from header parsing). Happy to tighten if the consensus is that whitespace should be rejected.

**Comment by hjmjohnson** [comment]
[dzenanz](https://github.com/dzenanz)
Approved these changes
